File: cpp/e2e_protection/basic_e2e.cpp
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Basic E2E Protection -- standalone CRC / counter / corruption detection demo.
 * Configuration is loaded from config/e2e_protection.yaml.
 */

#include <iostream>
#include <vector>
#include <string>

#include "config.h"
#include "e2e/e2e_protection.h"
#include "e2e/e2e_config.h"
#include "e2e/e2e_profiles/standard_profile.h"
#include "someip/message.h"
#include "common/result.h"

using namespace someip;
using namespace someip::e2e;

int main(int argc, char* argv[]) {
    std::string config_path;
    if (argc > 2 && std::string(argv[1]) == "--config") config_path = argv[2];

    examples::Config cfg("e2e_protection", config_path);

    auto data_id = static_cast<uint16_t>(cfg.get_int("e2e", "data_id", 0x1234));

    initialize_basic_profile();

    Message msg(MessageId(0x1234, 0x5678), RequestId(0x9ABC, 0xDEF0));
    msg.set_payload({0x01, 0x02, 0x03, 0x04, 0x05});

    E2EConfig e2e_cfg(data_id);
    e2e_cfg.enable_crc = true;
    e2e_cfg.enable_counter = true;
    e2e_cfg.enable_freshness = true;
    e2e_cfg.crc_type = 1;

    E2EProtection protection;

    std::cout << "=== SOME/IP Basic E2E Protection (C++) ===" << std::endl;

    Result result = protection.protect(msg, e2e_cfg);
    if (result != Result::SUCCESS) {
        std::cerr << "Failed to protect message" << std::endl;
        return 1;
    }
    std::cout << "Message protected successfully" << std::endl;
    std::cout << "E2E header present: " << (msg.has_e2e_header() ? "yes" : "no") << std::endl;

    std::vector<uint8_t> serialized = msg.serialize();
    std::cout << "Serialized message size: " << serialized.size() << " bytes" << std::endl;

    Message received_msg;
    if (!received_msg.deserialize(serialized, true)) {
        std::cerr << "Failed to deserialize message" << std::endl;
        return 1;
    }

    result = protection.validate(received_msg, e2e_cfg);
    if (result == Result::SUCCESS) {
        std::cout << "Message validated successfully" << std::endl;
        auto header_opt = received_msg.get_e2e_header();
        if (header_opt.has_value()) {
            const auto& header = header_opt.value();
            std::cout << "E2E Header:" << std::endl;
            std::cout << "  CRC: 0x" << std::hex << header.crc << std::endl;
            std::cout << "  Counter: " << std::dec << header.counter << std::endl;
            std::cout << "  Data ID: 0x" << std::hex << header.data_id << std::endl;
        }
    } else {
        std::cerr << "Message validation failed" << std::endl;
        return 1;
    }

    // Corruption detection
    std::cout << "\nTesting corruption detection..." << std::endl;
    received_msg.set_payload({0xFF, 0xFF});
    result = protection.validate(received_msg, e2e_cfg);
    if (result != Result::SUCCESS) {
        std::cout << "Corruption correctly detected" << std::endl;
    }

    std::cout << "\n=== E2E Protection Demo Complete ===" << std::endl;
    return 0;
}

File: python/e2e_protection/basic_e2e.py
#!/usr/bin/env python3
"""Basic E2E Protection -- standalone CRC / counter / corruption detection demo.

No transport is used -- this demonstrates E2E protection concepts.
Configuration from ``config/e2e_protection.yaml``.
"""

from __future__ import annotations

import struct
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from common.config import add_config_arg, load_config

from opensomeip.e2e import (
    E2ECheckStatus,
    E2EConfig,
    E2EProfile,
    E2EProfileId,
    E2EProtection,
    crc8,
    crc16,
    crc32,
)


class SimpleE2EProfile(E2EProfile):
    """A simple E2E profile that prepends CRC-16 + counter to the payload."""

    @property
    def profile_id(self) -> int:
        return E2EProfileId.PROFILE_01

    @property
    def header_size(self) -> int:
        return 4

    def protect(self, data: bytearray, counter: int) -> bytearray:
        crc = crc16(bytes(data))
        header = struct.pack("!HBB", crc, counter & 0xFF, 0x00)
        return bytearray(header) + data

    def check(self, data: bytes, counter: int) -> E2ECheckStatus:
        if len(data) < 4:
            return E2ECheckStatus.ERROR
        stored_crc, stored_counter, _ = struct.unpack("!HBB", data[:4])
        payload = data[4:]
        computed_crc = crc16(payload)
        if computed_crc != stored_crc:
            return E2ECheckStatus.ERROR
        if stored_counter != (counter & 0xFF):
            return E2ECheckStatus.WRONG_SEQUENCE
        return E2ECheckStatus.OK


def main() -> None:
    parser = add_config_arg()
    args = parser.parse_args()
    cfg = load_config("e2e_protection", args.config)

    data_id = cfg["e2e"]["data_id"]
    test_data = bytes(cfg["test_data"])

    print("=== SOME/IP Basic E2E Protection (Python) ===\n")

    # CRC demonstrations
    print("--- CRC Functions ---")
    print(f"Data: {test_data.hex()}")
    print(f"  CRC-8 (SAE-J1850): 0x{crc8(test_data):02X}")
    print(f"  CRC-16 (ITU-T):    0x{crc16(test_data):04X}")
    print(f"  CRC-32:            0x{crc32(test_data):08X}")

    # Sender/receiver E2E protection
    print("\n--- E2E Sender / Receiver ---")
    config = E2EConfig(
        profile_id=E2EProfileId.PROFILE_01,
        data_id=data_id,
        data_length=len(test_data),
    )
    profile = SimpleE2EProfile()

    sender = E2EProtection(config, profile=profile)
    receiver = E2EProtection(config, profile=profile)

    for i in range(3):
        protected = sender.protect(test_data)
        print(f"\n  Message {i}: {protected.hex()}")
        print(f"    Header: CRC=0x{protected[0]:02X}{protected[1]:02X} counter={protected[2]}")
        status = receiver.check(protected)
        print(f"    Check:  {status.name}")
        receiver._counter = (receiver._counter + 1) & 0xFF

    # Corruption detection
    print("\n--- Corruption Detection ---")
    protected = sender.protect(test_data)
    print(f"  Protected: {protected.hex()}")

    corrupted = bytearray(protected)
    corrupted[5] ^= 0xFF
    status = receiver.check(bytes(corrupted))
    print(f"  Corrupted payload -> {status.name}")

    corrupted2 = bytearray(protected)
    corrupted2[0] ^= 0xFF
    status2 = receiver.check(bytes(corrupted2))
    print(f"  Corrupted CRC     -> {status2.name}")

    print("\n=== E2E Protection Demo Complete ===")


if __name__ == "__main__":
    main()
